src/plots/cartesian/axes.js

```diff
@@ -3057,63 +3053,81 @@ axes.drawLabels = function(gd, ax, opts) {
+
+        var hideOverflow = ticklabeloverflow.indexOf('hide') !== -1;
+        // var pushOverflow = ticklabeloverflow.indexOf('push') !== -1;
+
+        var isX = ax._id.charAt(0) === 'x';
+        // div positions
+        var p0 = 0;
+        var p1 = isX ?
+            gd._fullLayout.width :
+            gd._fullLayout.height;
+
+        if(ticklabeloverflow.indexOf('domain') !== -1) {
+            // domain positions
             var rl = Lib.simpleMap(ax.range, ax.r2l);
+            p0 = ax.l2p(rl[0]) + ax._offset;
+            p1 = ax.l2p(rl[1]) + ax._offset;
+        }
 
-            // hide inside tick labels that go outside axis end points
-            var p0 = ax.l2p(rl[0]);
-            var p1 = ax.l2p(rl[1]);
+        var min = Math.min(p0, p1);
+        var max = Math.max(p0, p1);
 
-            var min = Math.min(p0, p1) + ax._offset;
-            var max = Math.max(p0, p1) + ax._offset;
+        var side = ax.side;
 
-            var side = ax.side;
-            var isX = ax._id.charAt(0) === 'x';
+        var visibleLabelMin = Infinity;
+        var visibleLabelMax = -Infinity;
 
-            var visibleLabelMin = Infinity;
-            var visibleLabelMax = -Infinity;
+        tickLabels.each(function(d) {
+            var thisLabel = d3.select(this);
+            var mathjaxGroup = thisLabel.select('.text-math-group');
 
-            tickLabels.each(function(d) {
-                var thisLabel = d3.select(this);
-                var mathjaxGroup = thisLabel.select('.text-math-group');
-
-                if(mathjaxGroup.empty()) {
-                    var bb = Drawing.bBox(thisLabel.node());
-                    var hide = false;
-                    if(isX) {
-                        if(bb.right > max) hide = true;
-                        else if(bb.left < min) hide = true;
-                    } else {
-                        if(bb.bottom > max) hide = true;
-                        else if(bb.top + (ax.tickangle ? 0 : d.fontSize / 4) < min) hide = true;
+            if(mathjaxGroup.empty()) {
+                var bb = Drawing.bBox(thisLabel.node());
+                var adjust = '';
+                if(isX) {
+                    if(bb.right > max) adjust += 'right';
+                    else if(bb.left < min) adjust += 'left';
+                } else {
+                    if(bb.bottom > max) adjust += 'top';
+                    else if(bb.top + (ax.tickangle ? 0 : d.fontSize / 4) < min) adjust += 'bottom';
+                }
+
+                var t = thisLabel.select('text');
+                if(adjust) {
+                    if(hideOverflow) t.style('opacity', 0); // hidden
+                    /*
+                    else if(pushOverflow) {
+                        if(adjust.indexOf('left') !== -1) t.attr('text-anchor', 'start');
+                        if(adjust.indexOf('right') !== -1) t.attr('text-anchor', 'end');
+                        // more TODO: https://github.com/plotly/plotly.js/issues/3292
```

Comment:
You could leave another branch open with this WIP for "push to ..." but I'd rather not have it commented out in the main codebase. Anyway the eventual solution is going to need to be a bit more complex than this: we'll want to push these labels only as far as necessary, and then if doing so causes them to intersect the next labels we should still hide them.

Reply:
Fixed by https://github.com/plotly/plotly.js/pull/5584/commits/c9eaff26eb1f31ebf10f1ba32c714f61e58106a7.